### ISSUE TITLE
BatchNormalization

### ISSUE THREAD

**Issue by HighPerformance01** [philipperemy/keras-tcn#44]
Hi philipperemy,

After I update tcn to the recent version, the training accuracy reduced tremendously (around from 0.98 to 0.17)

I then uncommented the following line of the code for BatchNormalization and things seems to be back to normal. So this seems to be an important part of tcn

x = BatchNormalization()(x)  # TODO should be WeightNorm here.

Is there a quick fix that going to happen soon?

Thank you

**Comment by HighPerformance01** [comment]
Also, there might be more than BatchNormalization line to the code. The old code (the old tcn code which is around 4 months old seems to be more reliable). However, this is still under testing.

**Comment by philipperemy** [comment]
@HighPerformance01 I saw that too but in the latest version, I commented out BatchNormalization.

Are you using the latest version? 

https://github.com/philipperemy/keras-tcn/blob/83c7ab5a6b84e61448d9d8f502836877cccc93f0/tcn/tcn.py#L35

Thanks for raising this

**Comment by philipperemy** [comment]
Also I pushed a new code that is very close to the paper, as requested by https://github.com/philipperemy/keras-tcn/issues/38

**Comment by philipperemy** [comment]
I could run every task (mnist, copy memory and adding problem) with the same accuracy and it seems like the new implementation was a bit better. So very looking forward to hearing more about your case.

**Comment by philipperemy** [comment]
But first please do:

```
pip install keras-tcn --upgrade
```

**Comment by philipperemy** [comment]
@HighPerformance01 any updates?

**Comment by HighPerformance01** [comment]
Thanks philipperemy for following up.

First, yes I am using the recent code but my data does not work until I uncommented BatchNormalization as follow:

x = BatchNormalization()(x) # TODO should be WeightNorm here.

So for my data, the training does not work without BatchNormalization as far as I can see.

But still, I see very little performance downgrade even after I uncommented BatchNormalization. My question here is there any additional change done to the code or params other than BatchNormalization.

I found the old code (four months old code) was performing better.

However, my data is very big and did not find optimal parameters yet, which may take maybe a few weeks more to conclude.

Thanks,

**Comment by philipperemy** [comment]
@HighPerformance01 that's interesting that it does not work without BatchNormalization. Your data must be big and your model deep.

Yes the model is a bit different. Its internal structure is slightly different. You can compare both models in terms of numbers of parameters and adjust the new model to have roughly the same number of parameters as the old one. It's possible that the new model has far less weights for the same configuration. So that could definitely explain the small performance drop.

Please keep me updated!

**Comment by philipperemy** [comment]
And sorry I misread "commented" with "uncommented" :)

**Comment by HighPerformance01** [comment]
Hi philipperemy,

Good news, I noticed a better accuracy using the latest today update tcn even better than any older tcn versions.

For BatchNormalization it seems it is related more to my data so I moved it to my code and used TCN function after that.

I find this repository very useful and I would only suggest not to limit the code to the paper but provide other extensions such as attention mechanism on top of tcn, in addition, to the paper implementation. Example:

tcn.py paper implementation.
tcn_attension.py example of extending tcn to use attention mechanism.

You may now close the question

Thank you  very much philipperemy for the useful repository

**Comment by philipperemy** [comment]
@HighPerformance01 ,

That's fantastic! I'll keep that in mind. 

Attention with TCN might actually be a very good combination!

Thanks again for your help,
